pandas/core/indexing.py

```diff
@@ -669,6 +670,56 @@ def _get_setitem_indexer(self, key):
 
         return self._convert_to_indexer(key, axis=0)
 
+    @final
+    def _maybe_mask_setitem_value(self, indexer, value):
+        """
+        If we have obj.iloc[mask] = arraylike and arraylike has the same
+        length as obj, we treat this as obj.iloc[mask] = arraylike[mask],
+        similar to Series.__setitem__.
+        """
+        # TODO: why do we only do this for loc, not iloc?
+        # ser = pd.Series(range(5))
+        # mask = np.array([True, False, True, False, True])
+        # ser[mask] = ser * 2  # <- works
+        # ser.loc[mask] = ser * 2  # <- works
+        # ser.iloc[mask] = ser * 2  # <- raises
+
+        if isinstance(indexer, tuple) and len(indexer) == 2:
+            pi, icols = indexer
+            ndim = getattr(value, "ndim", 0)
+            if ndim == 0:
+                pass
+            elif com.is_bool_indexer(pi) and len(value) == len(pi):
+                newkey = pi.nonzero()[0]
+
+                if is_scalar_indexer(icols, self.ndim - 1) and ndim == 1:
+                    # e.g. test_loc_setitem_boolean_mask_allfalse
+                    value = value[pi]
+                    indexer = (newkey, icols)
+
+                elif (
+                    isinstance(icols, np.ndarray)
+                    and icols.dtype.kind == "i"
+                    and len(icols) == 1
+                ):
+                    if ndim == 1:
+                        value = value[pi]
+                        indexer = (newkey, icols)
+                        # TODO: is it sketchy that icols is an ndarray and value
+                        #  is 1D?  i.e. should we require value to be 2D with
+                        #  value.shape[1] == 1?
+
+                    elif ndim == 2 and value.shape[1] == 1:
+                        if isinstance(value, ABCDataFrame):
+                            value = value.iloc[newkey]
```

Comment:
This causes bugs:

```
df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})

df.loc[[False, False, True], ["a"]] = pd.DataFrame({"a": [10, 20, 30]}, index=[2, 1, 0])
```

This returns 
```
    a  b
0   1  4
1   2  5
2  10  6
```
on main, which is correct and

```
     a    b
0  1.0  4.0
1  2.0  5.0
2  NaN  6.0
```

on this branch, which is wrong

Reply:
This has to be aligned before applying iloc

Reply:
good catch, will update+test

Reply:
so using align_series/align_frame fixes the examples you've provided, but breaks test_loc_setitem_boolean_mask_allfalse, test_setitem_loc_only_false_indexer_dtype_changed, test_loc_setitem_all_false_boolean_two_blocks, all cases where mask.sum() == 0.  ATM i can get the tests passing by putting in a check specific to that case, but its definitely a kludge

Reply:
Hm, it looks like the align functions can not deal with all False boolean indexers if I understand this correctly. This looks like a bug in there...

Reply:
If you set the indexer to something like ``indexer = (np.array([False, False, False]), 0)`` you get the identity.